ldmping/ldmping.1
.TH LDMPING 1 "2010-03-10"
.SH NAME
ldmping - check if ldm server is running
.SH "UNIX SYNOPSIS"
\fBldmping \fR[\fB\-x\fR] [\fB\-v\fR|\fB\-q\fR] [\fB\-l \fIlogdest\fR] 
.BI [\-t timeout]
.BI [\-i interval]
.BI [\-h remote]
.BI [remote ...]
.SH DESCRIPTION

This program sends a RPC NULLPROC request to a list of LDM servers
\fIremote ...\fR. The default behavior of the program changes depending
on whether the program is invoked interactively or not. (This is determined
by whether STDERR "isatty".)
.LP
This program is used as a simple check that an LDM server is running
on a given host and that the server is responding to RPC calls, or to
perform this check on a group of hosts.

.LP
.SH OPTIONS 
.TP 8
.B \-v
Verbose flag. 
If the verbose flag is on, a log message is emitted on success as well
as failure. If the program is invoked interactively, this is the default.
.TP 8
.B \-q
Quiet, turn off verbose flag. Useful to override the default.
.TP
.BI \-l " logdest"
Log to \fIlogdest\fP. One of \fB''\fP (system logging daemon), \fB'-'\fP
(standard error stream), or file \fIlogdest\fP. Default is the standard error
stream if the process has a controlling terminal (i.e., the process isn't a
daemon); otherwise, either the LDM log file or the system logging daemon
(execute this program with just the option \fB'-?'\fP to determine which).
.TP
.BI \-t " timeout"
Set the RPC timeout to this many seconds. The default is 10 seconds.
.TP
.BI \-i " interval"
Set the polling interval to this many seconds. If non zero, the program
will loop forever (until interrupted). If zero, the program tries each remote
once, exiting with status 0 if all check out okay, otherwise exiting with
status one (failure) when an error is encountered. The default polling interval
is 25 when interactive and 0 otherwise.
.TP
.BI \-h " remote"
Requests that  that the LDM server on \fIremote\fR be pinged. Multiple remotes
may be specified. A list of remotes may be also specified without using "-h".
If no remotes are specified, defaults to "localhost".
.TP
.B \-x
Requests that debugging information also be printed. (None in this program.)
.SH OUTPUT FORMAT
The report consists of 5 fields.
.LP
The first field is the state of the connection. The connection may be
seen as climbing through various states to complete a transaction.
The lowest state is NAMED, we have defined a service we wish to contact
(the ldm via TCP at a given \fIremotehost\fR).
The next state is ADDRESSED, meaning we were able to translate the
hostname into an IP address. (Sometimes RPC problems are really nameservice
problems.)
The third state, PMAP_CLNTED, indicates that the RPC portmapper is up on
the remote and the fourth, MAPPED, indicates that there is ldm service
(program 300029) registered.
The fifth state is H_CLNTED, indicating that we were able to make a
TCP connection to the port handed us by the RPC portmapper. Presumably this
is a connection to an ldm server.
Finally, if the remote service responds to the RPC NULLPROC within the
\fItimeout\fR interval, we mark it as RESPONDING.
.LP
The second field is the time elapsed in completing the call.
Expect that the Elapsed is greater during the initial connection than
during subsequent calls, since this time includes all the time taken
to climb thru the states.
.LP
The third field is the TCP port number (returned by the RPC portmapper).
.LP
The fourth field is the remote host you specified.
.LP
If there is an RPC problem, the RPC status will be printed at the end of
the line. Note that this status may apply to the RPC portmapper if the
state is lower than CONNECTED.
.SH "SEE ALSO"
ldmd(1), ingestd(1), ulog(3), rpcinfo(1)


CHANGE_LOG
6.13.0.0	2016-02-11T12:02:01-0700
    Logging:
        Re-wrote LDM logging because:
            1) It was the major impediment to correctly using the LDM;
            2) It was the major impediment to putting the LDM in a Docker
               container; and
            3) Some system logging daemons sometimes dropped LDM log messages.
        For backward-compatibility, the original logging component, "ulog",
        still exists and is unconditionally included in the LDM library,
        "libldm". Indeed, the original behavior of the LDM can be obtained by
        specifying the option "--with-ulog" to the configure(1) script. By
        default, however, the new logging component, named "slog" (for "simple
        logging"), will be built and used. While "slog" can be told to write to
        the system logging daemon, by default it will write directly to the LDM
        log file if the process doesn't have a controlling terminal (i.e., is a
        daemon) or to the standard error stream if the process does have a
        controlling terminal (i.e., is not a daemon). Like the original "ulog"
        component, "slog" also supports log file rotation via a
        crontab(1)-executed "ldmadmin newlog" command. See the installed HTML
        documentation "$HOME/share/doc/ldm/basics/logging-format.html" or the
        web page
        <http://www.unidata.ucar.edu/software/ldm/ldm-current/basics/logfile-format.html>
        for more information.

6.12.15.42	2016-01-12T10:40:24-0700
    Make test for C++11-compatible g++ more robust.

6.12.15.41	2016-01-12T09:28:01-0700
    Get "make distclean" to work on both Winters (with C++11 support) and Spock
    (without C++11 support).

6.12.15.40	2016-01-11T16:07:56-0700
    Removed mylog2ulog stuff from ulog makefile.
    
    Eliminated memory-leaks from components that read GEMPAK tables.

6.12.15.39	2015-11-16T16:04:41-0700
    mldm_sender_manager(3):
        Add "-y" and "-z" options to execvp(2) of mldm_sender(1).

    pqact.c:
        Added SIGHUP forwarding.

6.12.15.38	2015-11-03T13:07:13-0700
    ldmd.c:
        Corrected some Coverity-discovered lint.

6.12.15.37	2015-11-03T09:41:04-0700
    Increase resolution of printed product-creation timestamp from millisecond
    to microsecond.

6.12.15.36	2015-11-02T15:59:18-0700
    ldmadmin(1):
        Reduce overestimation of new queue capacities (i.e., data capacity,
        product capacity) and don't compute new capacities if the data is
        insufficient.
        
    logging:
        Added logging options for using microsecond resolution in timestamps and
        formatting them according to the ISO 8601 standard. Added respective
        options "-y" and "-z" to ldmd(1) and notifyme(1).

6.12.15.35	2015-10-28T15:29:59-0600
    Create mcast_test(1) for testing LDM-7 on GENI.

6.12.15.34	2015-10-26T16:41:04-0600
    PerProdNotifier:
        Fix setting of product-information by notify_of_bop().

6.12.15.33	2015-10-23T15:57:05-0600
    mldm_sender():
        Moved addition of product-index-to-product-signature entry to before the
        sending of the product because the receiving LDM-7 would sometimes
        request a data-product before the mapping was entered resulting in the
        product not being found.

6.12.15.32	2015-10-22T19:32:42-0600
    Improve receiver session-memory API.

6.12.15.31	2015-10-20T17:11:53-0600
    down7.c:
        Correct location of down7_incNumProds().

        Add mutex destruction to down7_free().

    mldm_sender:
        Add opening of product-queue in thread-safe mode to prevent corruption
        of the regions-in-use list by multiple thread access.

    pq.c:
        Add unlocking of mutex in pq_sequence() before calling `ifMatch()`
        to avoid deadlock in calling a foreign function.

        Add mutex destruction to freeing.

6.12.15.30	2015-10-19T16:23:06-0600
    noaaportIngester(1):
        Added "+000 001 008 010 Total precipitation" to GEMPAK tables.

    Multicast:
        Add feature to lock products until VCMTP sender releases them.

        Modify VCMTP port numbers to match WireShark plugin.

        Improve log messages.

    pq(3):
        Make pq_release() thread-safe.

6.12.15.29	2015-10-08T18:03:25-0600
    Add diagnostic print statements and modify up7-down7_test to try
    diagnosing product-queue corruption seen by Shawn chen.

6.12.15.28	2015-10-06T16:03:42-0600
    pq(3):
        Modified fb_get() to log fblk metrics when it fails to help diagnose
        multicast problem.

6.12.15.27	2015-10-06T12:01:45-0600
    pqutil(1):
        Removed minor lint revealed by Coverity Scan.

    pq(3):
        Use correct product size in pqe_insert() and pq_insertNoSig().

    mcast_lib(3):
        Adapt to new checks by pqe_insert().

6.12.15.26	2015-10-06T09:48:58-0600
    pqact(1):
        Correct the setting of the product-queue cursor based on the last
        successfully-processed data-product from the previous session.
        Before this, pqact(1) might miss a product due to a restart of the LDM.
        This was most apparent in processing NEXRAD-2 data-products. Thanks
        Ryan.

    Product-Queue (pq(3)):
        Refactored pqe_insert() for better error-handling and return status.

        Harmonized return code names by adding PQ_DUP and PQ_BIG and making
        PQUEUE_DUP and PQUEUE_BIG aliases.

6.12.15.25	2015-10-01T15:28:56-0600
    Remove ignoring of SIGCONT from mldm_sender_manager because it's not needed.

    Generalize TCP keep-alive in TcpSend.cpp.

6.12.15.24	2015-09-30T15:40:21-0600
    Reduce number of "make distcheck" to just one in CI-commit.sh.

6.12.15.23	2015-09-30T14:33:58-0600
    Replace "$@" with just $@ in deploy.sh.

6.12.15.22	2015-09-30T13:32:25-0600
    Replace ${configOpts} with "$@" in deploy.sh to accommodate argument with
    space.

6.12.15.21	2015-09-29T16:57:17-0600
    Corrected setting of configure(1) options in delivery/deploy.sh.

6.12.15.20	2015-09-29T15:03:45-0600

ldmping/ldmping.c
/*
 *   Copyright 1993, University Corporation for Atmospheric Research
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: ldmping.c,v 1.20.18.3 2007/04/05 21:41:58 steve Exp $ */

/* 
 * pings remote host
 */

#include "config.h"

#include <errno.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "h_clnt.h"
#include "ldm5.h"
#include "ldm5_clnt.h"
#include "log.h"


#define DEFAULT_INTERVAL 25
#define DEFAULT_TIMEO 10


/*
 * ping the remote
 */
static enum clnt_stat
check_hstat(hcp, timeout)
h_clnt *hcp ;
int timeout ;
{
        return nullproc5(hcp, timeout) ;
}


static void
print_label()
{
        log_info("%10s %10s %4s   %-21s %s\n",
                        "State",
                        "Elapsed",
                        "Port",
                        "Remote_Host",
                        "rpc_stat"
                        ) ;
        return ;
}


static void
print_hstat(hcp)
h_clnt *hcp ;
{
        if(hcp->state == RESPONDING)
                log_info("%10s %3ld.%06ld %4d   %-11s  %s\n",
                        s_remote_state(hcp->state),
                        hcp->elapsed.tv_sec, hcp->elapsed.tv_usec,
                        hcp->port,
                        hcp->remote,
                        s_hclnt_sperrno(hcp)
                        ) ;
        else
                log_error("%10s %3ld.%06ld %4d   %-11s  %s\n",
                        s_remote_state(hcp->state),
                        hcp->elapsed.tv_sec, hcp->elapsed.tv_usec,
                        hcp->port,
                        hcp->remote,
                        s_hclnt_sperrno(hcp)
                        ) ;
        return ;
}


static void
usage(av0)
char *av0 ; /*  id string */
{
        (void)fprintf(stderr,
"Usage: %s [options] [remote ...] \t\nOptions:\n", av0);
        (void)fprintf(stderr,
"\t-v           Verbose (default if interactive)\n") ;
        (void)fprintf(stderr,
"\t-q           Quiet (to shut up when interactive)\n") ;
        (void)fprintf(stderr,
"\t-x           Debug mode\n") ;
        (void)fprintf(stderr,
"\t-l dest      Log to `dest`. One of: \"\" (system logging daemon), \"-\"\n"
"\t             (standard error), or file `dest`. Default is \"%s\"\n",
                log_get_default_destination());
        (void)fprintf(stderr,
"\t                 else uses syslogd)\n") ;
        (void)fprintf(stderr,
"\t-t timeout   set RPC timeout to \"timeout\" seconds (default %d)\n",
                        DEFAULT_TIMEO) ;
        fprintf(stderr,
"\t-i interval  Poll after \"interval\" secs (default %d when interactive,\n",
                DEFAULT_INTERVAL) ;
        (void)fprintf(stderr,
"\t                 0 => one trip otherwise)\n") ;
        (void)fprintf(stderr,
"\t-h remote    \"remote\" host to ping (default is localhost)\n") ;
        exit(1);
}


int main(ac,av)
int ac ;
char *av[] ;
{
        int verbose = 0 ;
        int interval = 0 ;
        int timeo = DEFAULT_TIMEO ; 
        int logoptions = (LOG_CONS|LOG_PID) ;
        int nremotes = 0 ;
#define MAX_REMOTES 14 /* 2 * MAX_REMOTES + 3 < max_open_file_descriptors */
        h_clnt stats[MAX_REMOTES + 1] ;
        h_clnt *sp ;
        unsigned        port = 0;
        unsigned        logmask = LOG_UPTO(LOG_LEVEL_NOTICE);

        /*
         * initialize logger
         */
        (void)log_init(av[0]);

        if(isatty(fileno(stderr)))
        {
                /* set interactive defaults */
                verbose = !0 ;
                interval = DEFAULT_INTERVAL ;
        }

        {
        extern int optind;
        extern int opterr;
        extern char *optarg;
        int ch;

        opterr = 1;

        while ((ch = getopt(ac, av, "vxl:t:h:P:qi:")) != EOF)
                switch (ch) {
                case 'v':
                        if (!log_is_enabled_info)
                            (void)log_set_level(LOG_LEVEL_INFO);
                        verbose = !0 ;
                        break;
                case 'q':
                        verbose = 0 ;
                        break ;
                case 'x':
                        (void)log_set_level(LOG_LEVEL_DEBUG);
                        break;
                case 'l':
                        (void)log_set_destination(optarg);
                        break;
                case 'h':
                        if(nremotes > MAX_REMOTES)
                        {
                                fprintf(stderr,
                                        "Can't handle more than %d remotes\n", MAX_REMOTES) ;
                                break ;
                        }
                        init_h_clnt(&stats[nremotes++], optarg,
                                LDMPROG, FIVE, IPPROTO_TCP) ;
                        break ;
                case 'P': {
                    char*       suffix = "";
                    long        p;

                    errno = 0;
                    p = strtol(optarg, &suffix, 0);

                    if (0 != errno || 0 != *suffix ||
                        0 >= p || 0xffff < p) {

                        (void)fprintf(stderr, "%s: invalid port %s\n",
                             av[0], optarg);
                        usage(av[0]);   
                    }
                    else {
                        port = p;
                    }

                    break;
                }
                case 't':
                        timeo = atoi(optarg) ;
                        if(timeo == 0 && *optarg != '0')
                        {
                                fprintf(stderr, "%s: invalid timeout %s", av[0], optarg) ;
                                usage(av[0]) ;  
                        }
                        break;
                case 'i':
                        interval = atoi(optarg) ;
                        if(interval == 0 && *optarg != '0')
                        {
                                fprintf(stderr, "%s: invalid interval %s", av[0], optarg) ;
                                        usage(av[0]) ;
                        }
                        break ;
                case '?':
                        usage(av[0]);
                        break;
                }

        for(; nremotes < MAX_REMOTES && optind < ac ; nremotes++, optind++)
        {
                init_h_clnt(&stats[nremotes], av[optind],
                        LDMPROG, FIVE, IPPROTO_TCP) ;
        }
        if(ac - optind > 0)
        {
                fprintf(stderr,
                        "Can't handle more than %d remotes\n", MAX_REMOTES) ;
        }
        if(nremotes == 0)
        {
                init_h_clnt(&stats[nremotes++], "localhost",
                        LDMPROG, FIVE, IPPROTO_TCP) ;
        }
        stats[nremotes].state = H_NONE ; /* terminator */
        }

        /*
         * set up signal handlers
         */
        (void) signal(SIGPIPE, SIG_IGN) ;

        if(verbose)
                print_label() ;

        while(1)
        {
                for(sp = stats ; sp->state != H_NONE ; sp++)
                {
                        check_hstat(sp, timeo) ; 
                        /* if not verbose, only report "significant" stuff */
                        if(verbose || sp->elapsed.tv_sec > 1 || sp->state != RESPONDING)
                                print_hstat(sp) ;
                        if(interval == 0 && sp->state != RESPONDING)
                                exit(1) ;
                }
                if(interval == 0)
                        break ; 
                sleep(interval) ;
        }

        exit(0) ;
}
